homeassistant/components/nederlandse_spoorwegen/sensor.py

```diff
@@ -2,289 +2,215 @@
-    }
-)
+from homeassistant.helpers.device_registry import DeviceInfo
+from homeassistant.helpers.entity import EntityCategory  # type: ignore[attr-defined]
+from homeassistant.helpers.entity_platform import AddConfigEntryEntitiesCallback
+from homeassistant.helpers.update_coordinator import CoordinatorEntity
 
-ROUTES_SCHEMA = vol.All(cv.ensure_list, [ROUTE_SCHEMA])
+from . import NSConfigEntry
+from .const import CONF_FROM, CONF_TO, CONF_VIA, DOMAIN
+from .coordinator import NSDataUpdateCoordinator
 
-PLATFORM_SCHEMA = SENSOR_PLATFORM_SCHEMA.extend(
-    {vol.Required(CONF_API_KEY): cv.string, vol.Optional(CONF_ROUTES): ROUTES_SCHEMA}
-)
+_LOGGER = logging.getLogger(__name__)
 
 
-def setup_platform(
+async def async_setup_entry(
     hass: HomeAssistant,
-    config: ConfigType,
-    add_entities: AddEntitiesCallback,
-    discovery_info: DiscoveryInfoType | None = None,
+    entry: NSConfigEntry,
+    async_add_entities: AddConfigEntryEntitiesCallback,
 ) -> None:
-    """Set up the departure sensor."""
-
-    nsapi = ns_api.NSAPI(config[CONF_API_KEY])
-
-    try:
-        stations = nsapi.get_stations()
-    except (
-        requests.exceptions.ConnectionError,
-        requests.exceptions.HTTPError,
-    ) as error:
-        _LOGGER.error("Could not connect to the internet: %s", error)
-        raise PlatformNotReady from error
-    except RequestParametersError as error:
-        _LOGGER.error("Could not fetch stations, please check configuration: %s", error)
+    """Set up NS sensors from a config entry."""
+    coordinator = entry.runtime_data.get("coordinator")
+    if coordinator is None:
+        _LOGGER.error("Coordinator not found in runtime_data for NS integration")
         return
-
-    sensors = []
-    for departure in config.get(CONF_ROUTES, {}):
-        if not valid_stations(
-            stations,
-            [departure.get(CONF_FROM), departure.get(CONF_VIA), departure.get(CONF_TO)],
-        ):
-            continue
-        sensors.append(
-            NSDepartureSensor(
-                nsapi,
-                departure.get(CONF_NAME),
-                departure.get(CONF_FROM),
-                departure.get(CONF_TO),
-                departure.get(CONF_VIA),
-                departure.get(CONF_TIME),
+    _LOGGER.debug(
+        "NS sensor setup: coordinator=%s, entry_id=%s", coordinator, entry.entry_id
+    )
+
+    # Always create the service sensor
+    entities: list[SensorEntity] = [NSServiceSensor(coordinator, entry)]
+
+    # Create trip sensors for each route
+    if coordinator.data and "routes" in coordinator.data:
+        for route_key, route_data in coordinator.data["routes"].items():
+            route = route_data["route"]
+            entities.append(
+                NSTripSensor(
+                    coordinator,
+                    entry,
+                    route,
+                    route_key,
+                )
             )
-        )
-    add_entities(sensors, True)
```

Comment:
We should import the old configuration so the user only has to remove their old config and they'd be ready.